BAW mesh loader the 2nd PR

Crisspl opened the pull request:
While coding I figured out we don't need `getNeededDeps` if we have `instantiateEmpty`. Also we don't need two stacks (in loader) about which we were talking about. Now there's just vector with blobs' data, map<handle, object ptr> and two loops performing actual loading.  
Will merge after CR.

Left todo:  
* There's some bug in loading/exporting skinned meshes. Working on it, but couldn't find anything wrong so far. It works (doesn't crash), but can't see anything or.. something weird happens. Please run the modified 7th example so you can see the effect.
* Bit fields in `SMaterial`

Crisspl committed:
Let me go!

Crisspl committed:
Merge branch 'master' into master

Crisspl committed:
CBAWMeshWriter: skinned meshes plus some docs

Crisspl committed:
Variable size blobs support, some docs fixes

Crisspl committed:
oops

Crisspl committed:
BlobHeaderV1::finalize()

Crisspl committed:
BAW mesh loader

Working but need some more testing (especially skinned meshes - not tested at all). BAW mesh writer needs testing for skinned meshes as well.

Crisspl committed:
Modified MeshLoaders example to test BAW mesh loader

Crisspl committed:
dont look at this

Crisspl committed:
Fixed an object multiplicating bug in loader, this was serious one..

Crisspl committed:
Some cleaning up

Crisspl committed:
BAW mesh loader: Verifying file format

devshgraphicsprogramming committed:
Look at the magic of the Curiously Recurring Template Parameter

devshgraphicsprogramming committed:
Yeah, createTryOnStack for all VariableSizeBlobs

Lets keep the API consistent and create all blobs and pull all their data in their respective constructors (private for VariableSizeBlob inheritors).

Also validatation of blobs before load, more needs to be added, see the comments in the code.

Crisspl committed:
BAW mesh loader: Changes after CR

Done: moved loading blobs to member functions of blob-structs; made an abstraction between blob version and loading; reading file blob by blob (not all at once); file validation etc.; eliminated recursive blobs loading; eliminated DRY violations; reworked CFinalBoneHierarchy loading

Crisspl committed:
Changed BAWFile names (upper case); TypedBlob base; *Serializable base classes; Ctor for FBH blob

We have to talk about mesh/skinned-mesh inheritance from VariableSizeBlobSerializable (ISkinnedMesh inherits IMesh), same with mesh buffers

Crisspl committed:
CBlobsLoadingManager: map -> switch; Reworked baw mesh loader a little along with CBlobsLoadingManager

Crisspl committed:
Part of the docs

Crisspl committed:
Merge branch 'master' into master

Crisspl committed:
Update project files for Windows build

Crisspl committed:
BlobSerializable

devshgraphicsprogramming commented on include/CBAWFile.h:
```diff
@@ -179,63 +209,66 @@ namespace core
+	struct MeshDataFormatDescBlobV0 : TypedBlob<MeshDataFormatDescBlobV0, scene::IMeshDataFormatDesc<core::ICPUBuffer> >, FixedSizeBlob<MeshDataFormatDescBlobV0, scene::IMeshDataFormatDesc<core::ICPUBuffer> >
 	{
+	private:
+		enum { VERTEX_ATTRIB_CNT = 16 };
```
can we make it equal to scene::EVAI_COUNT ?

or do we run into circular refs?

Crisspl replied:
> or do we run into circular refs?  
  
Yes, I had to forward declare every class used by blobs to make them able to inherit from classes from CBAWFile.h. So there's the best solution I came up with. I put assert (or rather `_IRR_DEBUG_BREAK_IF` so we will know 
when `EVAI_COUNT` change https://github.com/buildaworldnet/IrrlichtBAW/pull/36/commits/22ddd15def5d4041b3480288da97e5c6d7859784#diff-b7b624bc03dce20e1e4c837d54b34971R103

devshgraphicsprogramming replied:
good solution

devshgraphicsprogramming commented on include/IMeshBuffer.h:
```diff
@@ -161,7 +162,7 @@ GL_INVALID_OPERATION is generated by glVertexAttribPointer if size is GL_BGRA an
 
 	template <class T>
-	class IMeshDataFormatDesc : public virtual IReferenceCounted
+	class IMeshDataFormatDesc : public virtual IReferenceCounted, public core::BlobSerializable<IMeshDataFormatDesc<core::ICPUBuffer> >
```
this makes the entire template class serializable, not just ICPUMeshDataFormatDesc (which is a typedef for IMeshDataFormatDesc<ICPUBuffer>)

will have to convert the ICPUMeshDataFormatDesc typedef into a class inheriting from IMeshDataFormatDesc<ICPUBuffer> and BlobSerializable

Crisspl replied:
Yea, I know it's a todo, got it in mind

devshgraphicsprogramming commented on include/ICPUBuffer.h:
```diff
@@ -37,10 +37,18 @@ class ICPUBuffer : public IBuffer
             size = sizeInBytes;
         }
 
+		//! Returns buffer's type which in this case is always equal to `E_BUFFER_TYPE::EBT_UNSPECIFIED_BUFFER`.
```
we initially wanted to support buffer typing (like in DX10) but OpenGL  4.0+ brought ARB_buffer_storage which makes the buffer type irrelevant, since you explicitly state if the buffer should be on the GPU or not.
Aside from that OpenGL never had strict buffer typing, the buffer type was more of a hint as to whether to store the buffer in RAM,DMA RAM or VRAM.
So this E_BUFFER_TYPE enum is deprecated.

devshgraphicsprogramming commented on include/ICPUBuffer.h:
```diff
@@ -37,10 +37,18 @@ class ICPUBuffer : public IBuffer
 
+		//! Reallocates internal data. Invalidate any sizes, pointers etc. returned before!
+		/** @param newSize New size of memory.
+		@param forceRetentionOfData Doesn't matter.
```
Doesn't matter right now, but if we switch to a memory allocator that does not offer a free `realloc` then this will necessitate a memcpy.

Crisspl replied:
It's one of things I wasn't sure about meaning. I'll ask you about this later anyway :)

devshgraphicsprogramming commented on include/IFrameBuffer.h:
```diff
@@ -50,12 +53,31 @@ class IRenderBuffer;
+		/** @param attachmentPoint Identifies attchment point.
+		@param tex Texture being attached.
+		@param mipMapLayer
+		@param layer
```
layer of the framebuffer to attach to, for explanation see the OpenGL Wiki for the topic of "layered rendering"
values >= mean that a particular `layer` of a 2D or cubemap texture array, or 3D texture is attached to the FrameBuffer.
value <0 means the entire 3D texture or, 2D texture or cubemap array is bound making the FrameBuffer layered, and enabling you to use gl_Layer for layered rendering.

Look at example 15 to see layered rendering used on a cubemap to render point shadows in a single pass.

devshgraphicsprogramming commented on include/IFrameBuffer.h:
```diff
@@ -50,12 +53,31 @@ class IRenderBuffer;
+		@param tex Texture being attached.
+		@param mipMapLayer
+		@param layer
+		@returns Whether attachment operation went succesfully.
```
Only returns if the attachment has been attached (glFramebuffer* called) and the FBO is queued up for rebinding.
A return value of TRUE does not mean that the attachment color format is renderable or that the combination of attachments is valid.
Only after rebindRevalidate() is called by the driver internally or by the user manually do the attachments drawn into by the FrameBuffer change.

devshgraphicsprogramming commented on include/IFrameBuffer.h:
```diff
@@ -50,12 +53,31 @@ class IRenderBuffer;
+		//! Rebinds frame buffer object.
+		/** @returns whether FBO is bound or was successfully rebound.
+		*/
         virtual bool rebindRevalidate() = 0;
```
This is a general function present in all our video namespace objects which reference/use resizeable resources such as textures and buffers.
ARB_buffer_storage and ARB_texture_storage remove the ability to respecify storage for OpenGL objects (essentially no GPU realloc) so to resize objects we need to destroy the old and create a new object, this means that the OpenGL handles (object IDs) change so the new object needs to be bound in place of the old (texture to an FBO, buffer to a VAO).
Whenever a bound object gets modified this way an atomic counter increases and this triggers the rebind* function to do more than just return `true`.

The driver always calls these functions internally before using the object with a rebind*() function.

However for the FrameBuffer you can call it manually after changing attachments as it will also call `glCheckFramebufferStatus` which will make the rebindRevalidate() function return false if the color formats you are trying to render to are invalid (like RGB9_E5) or if a certain combination of attachments is invalid (try specifying the DEPTH and STENCIL attachments to be different with any Nvidia GPU).

This function does nothing if no changes to bound objects have occurred.

devshgraphicsprogramming commented on include/IGPUMappedBuffer.h:
```diff
@@ -17,10 +17,13 @@ namespace video
+		//! Gets internal pointer.
+        /** WARNING: RESIZE will invalidate pointer!
+        WARNING: NEED TO FENCE BEFORE USE!
+		@returns Internal pointer. */
```
Its best you use a GPU Fence to ensure any operations which you've queued up that are writing to this buffer or reading from it have completed before you start using this pointer, otherwise you will have a race condition.

devshgraphicsprogramming commented on include/IGPUMappedBuffer.h:
```diff
@@ -17,10 +17,13 @@ namespace video
+		@returns Internal pointer. */
         virtual void* getPointer() = 0;
 
+		//! @returns Always true.
```
We only support persistently mapped buffers with ARB_buffer_storage.
Its almost the fastest across all cards, and is more in line with what Vulkan has.
Please don't ask us to support Buffer Orpahning, or Map/Unmap.

devshgraphicsprogramming commented on include/SCollisionEngine.h:
```diff
@@ -15,13 +15,22 @@ class SCollisionEngine
 
 		//! Returns a 3d ray which would go through the 2d screen coodinates.
+		/**
+		@param[out] origin Attachment point of the output ray
```
start point

devshgraphicsprogramming commented on include/SCollisionEngine.h:
```diff
@@ -55,6 +64,13 @@ class SCollisionEngine
+		@param driver Driver
+		@param camera Camera on which calculations will depend
+		@param iseViewPort Whether to use viewport or current render target's size
+		@returns {-100000, -100000} (minus ten thousand) if an error occured
```
if the point is behind the camera

devshgraphicsprogramming commented on include/coreutil.h:
```diff
@@ -393,6 +478,7 @@ inline void XXHash_256(const void* input, size_t len, uint64_t* out)
     out[3] += v4;
 }
 
+//! Utility class easing the process of finding memory leaks. Usable only in debug build. Thread-safe.
```
Doesn't have a Windows implementation yet.

Crisspl replied:
It's mostly not documented. I'll have to ask what `getBackTrace` actually do because i couldn't figure out.

Crisspl committed:
BAW mesh loader: load only necessary blobs

Crisspl committed:
[wip] Docs fixes

Crisspl committed:
baw mesh loader optimizations

Each needed blob loaded once

Crisspl committed:
Patched a few major bugs

1) not rendering texture (bad cast); 2) crash while loading skinned meshes (calling primary template definition instead of specialization of func calculating blob's size); 3) broken animations in bone hierarchy (not all animations were exported/loaded)

Crisspl committed:
Reworked BlobSerializable class; Renamed blobsLoading.cpp file

Crisspl committed:
Lil bug

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshFileLoader.h:
```diff
@@ -32,12 +32,12 @@ class CBAWMeshFileLoader : public IMeshLoader
-			validated = true;
-			return validated ? true : (heapBlob && header->validate(heapBlob)); 
+			validated = false;
+			return validated ? true : (validated = (heapBlob && header->validate(heapBlob)));
```
careful about short-circuit evaluation

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -186,7 +186,7 @@ namespace irr {
 			if (_mesh)
 			{
-				skinnedMesh = _mesh->getMeshType()!=EMT_ANIMATED_SKINNED ? NULL:reinterpret_cast<ICPUSkinnedMesh*>(_mesh); //ICPUSkinnedMesh is a direct non-virtual inheritor
+				skinnedMesh = _mesh->getMeshType()!=EMT_ANIMATED_SKINNED ? NULL:dynamic_cast<ICPUSkinnedMesh*>(_mesh); //ICPUSkinnedMesh is a direct non-virtual inheritor
```
do we need dynamic_cast instead of static_cast ?

(reinterpret is always wrong)
